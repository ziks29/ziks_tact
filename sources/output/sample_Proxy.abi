{"name":"Proxy","types":[{"name":"StateInit","header":null,"fields":[{"name":"code","type":{"kind":"simple","type":"cell","optional":false}},{"name":"data","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"Context","header":null,"fields":[{"name":"bounced","type":{"kind":"simple","type":"bool","optional":false}},{"name":"sender","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"raw","type":{"kind":"simple","type":"slice","optional":false}}]},{"name":"SendParameters","header":null,"fields":[{"name":"bounce","type":{"kind":"simple","type":"bool","optional":false}},{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"mode","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"body","type":{"kind":"simple","type":"cell","optional":true}},{"name":"code","type":{"kind":"simple","type":"cell","optional":true}},{"name":"data","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"ChangeOwner","header":2174598809,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"ChangeOwnerOk","header":846932810,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"Deploy","header":2490013878,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DeployOk","header":2952335191,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"FactoryDeploy","header":1829761339,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"cashback","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"ProxyMessage","header":2376135367,"fields":[{"name":"str","type":{"kind":"simple","type":"string","optional":false}},{"name":"to","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"ChangeExampleOwner","header":3797952890,"fields":[{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"LastMessage","header":null,"fields":[{"name":"lastMessage","type":{"kind":"simple","type":"string","optional":true}},{"name":"lastSender","type":{"kind":"simple","type":"address","optional":true}},{"name":"lastReciver","type":{"kind":"simple","type":"address","optional":true}}]}],"receivers":[{"receiver":"internal","message":{"kind":"text","text":"Repeat"}},{"receiver":"internal","message":{"kind":"text","text":"Return"}},{"receiver":"internal","message":{"kind":"text"}},{"receiver":"internal","message":{"kind":"typed","type":"ProxyMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"ChangeExampleOwner"}},{"receiver":"internal","message":{"kind":"typed","type":"Deploy"}}],"getters":[{"name":"owner","arguments":[],"returnType":{"kind":"simple","type":"address","optional":false}},{"name":"get_count","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"get_last","arguments":[],"returnType":{"kind":"simple","type":"LastMessage","optional":false}}],"errors":{"2":{"message":"Stack undeflow"},"3":{"message":"Stack overflow"},"4":{"message":"Integer overflow"},"5":{"message":"Integer out of expected range"},"6":{"message":"Invalid opcode"},"7":{"message":"Type check error"},"8":{"message":"Cell overflow"},"9":{"message":"Cell underflow"},"10":{"message":"Dictionary error"},"13":{"message":"Out of gas error"},"32":{"message":"Method ID not found"},"34":{"message":"Action is invalid or not supported"},"37":{"message":"Not enough TON"},"38":{"message":"Not enough extra-currencies"},"128":{"message":"Null reference exception"},"129":{"message":"Invalid serialization prefix"},"130":{"message":"Invalid incoming message"},"131":{"message":"Constraints error"},"132":{"message":"Access denied"},"133":{"message":"Contract stopped"},"134":{"message":"Invalid argument"},"135":{"message":"Code of a contract was not found"},"136":{"message":"Invalid address"},"137":{"message":"Masterchain support is not enabled for this contract"},"27831":{"message":"Only owner can call this function"}},"interfaces":["org.ton.introspection.v0","org.ton.abi.ipfs.v0","org.ton.deploy.lazy.v0","org.ton.chain.workchain.v0","org.ton.ownable"]}