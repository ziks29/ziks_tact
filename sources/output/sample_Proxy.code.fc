#pragma version =0.4.3;
#pragma allow-post-modification;
#pragma compute-asm-ltr;

#include "sample_Proxy.headers.fc";
#include "sample_Proxy.stdlib.fc";
#include "sample_Proxy.storage.fc";

;;
;; Contract Proxy functions
;;

(slice, int, slice, slice, slice) $Proxy$_contract_init(slice $owner) impure inline_ref {
    var (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)) = (null(), 0, null(), null(), null());
    $self'owner = $owner;
    return ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver);
}

((slice, int, slice, slice, slice), ()) $Proxy$_fun_storeStats((slice, int, slice, slice, slice) $self, slice $sender, slice $receiver, slice $message) impure inline_ref {
    var (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)) = $self;
    $self'count = ($self'count + 1);
    $self'lastSender = $sender;
    $self'lastReciver = $receiver;
    $self'lastMessage = $message;
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), ());
}

((slice, int, slice, slice, slice), ()) $Proxy$_fun_requireOwner((slice, int, slice, slice, slice) $self) impure inline_ref {
    var (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)) = $self;
    throw_unless(132, ( __tact_slice_eq_bits($self'owner, __tact_context_get_sender()) ));
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), ());
}

((slice, int, slice, slice, slice), slice) $Proxy$_fun_owner((slice, int, slice, slice, slice) $self) impure inline_ref {
    var (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)) = $self;
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), $self'owner);
}

((slice, int, slice, slice, slice), int) $Proxy$_fun_get_count((slice, int, slice, slice, slice) $self) impure inline_ref {
    var (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)) = $self;
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), $self'count);
}

((slice, int, slice, slice, slice), (slice, slice, slice)) $Proxy$_fun_get_last((slice, int, slice, slice, slice) $self) impure inline_ref {
    var (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)) = $self;
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), $LastMessage$_constructor_lastMessage_lastSender_lastReciver($self'lastMessage, $self'lastSender, $self'lastReciver));
}

;;
;; Receivers of a Contract Proxy
;;

((slice, int, slice, slice, slice), ()) $Proxy$_internal_text_561db9191cda1922faf6066ed24edc68166b3930bfeb59ba7fc844712e788f75((slice, int, slice, slice, slice) $self) impure inline {
    var ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver) = $self;
    ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)~$Proxy$_fun_requireOwner();
    $global_send($SendParameters$_constructor_to_value_mode_body(__tact_not_null($self'lastReciver), 0, 64, $String$_fun_asComment(__tact_not_null($self'lastMessage))));
    $self'count = ($self'count + 1);
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), ());
}

((slice, int, slice, slice, slice), ()) $Proxy$_internal_text_6ef41a1603279cbd03396d3a3593d3f465c968554125a8b39067ebfc8b060d65((slice, int, slice, slice, slice) $self) impure inline {
    var ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver) = $self;
    var ($ctx'bounced, $ctx'sender, $ctx'value, $ctx'raw) = __tact_context_get();
    throw_unless(27831, ( __tact_slice_eq_bits($self'owner, $ctx'sender) ));
    $global_send($SendParameters$_constructor_to_value_mode_body(__tact_not_null($self'lastSender), 0, 64, $String$_fun_asComment(__tact_not_null($self'lastMessage))));
    $self'count = ($self'count + 1);
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), ());
}

((slice, int, slice, slice, slice), ()) $Proxy$_internal_any_text((slice, int, slice, slice, slice) $self, slice $str) impure inline {
    var ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver) = $self;
    $global_send($SendParameters$_constructor_to_value_mode_body($self'owner, 0, 64, $String$_fun_asComment($str)));
    var ($ctx'bounced, $ctx'sender, $ctx'value, $ctx'raw) = __tact_context_get();
    ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)~$Proxy$_fun_storeStats($ctx'sender, $self'owner, $str);
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), ());
}

(((slice, int, slice, slice, slice)), ()) $Proxy$_internal_binary_ProxyMessage((slice, int, slice, slice, slice) $self, (slice, slice) $msg) impure inline {
    var ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver) = $self;
    var ($msg'str, $msg'to) = $msg;
    $global_send($SendParameters$_constructor_to_value_mode_body($msg'to, 0, 64, $String$_fun_asComment($msg'str)));
    var ($ctx'bounced, $ctx'sender, $ctx'value, $ctx'raw) = __tact_context_get();
    ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)~$Proxy$_fun_storeStats($ctx'sender, $msg'to, $msg'str);
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), ());
}

(((slice, int, slice, slice, slice)), ()) $Proxy$_internal_binary_ChangeExampleOwner((slice, int, slice, slice, slice) $self, (slice) $msg) impure inline {
    var ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver) = $self;
    var ($msg'newOwner) = $msg;
    ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)~$Proxy$_fun_requireOwner();
    $self'owner = $msg'newOwner;
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), ());
}

(((slice, int, slice, slice, slice)), ()) $Proxy$_internal_binary_Deploy((slice, int, slice, slice, slice) $self, (int) $deploy) impure inline {
    var ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver) = $self;
    var ($deploy'queryId) = $deploy;
    ($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver)~$Proxy$_fun_notify($DeployOk$_store_cell($DeployOk$_constructor_queryId($deploy'queryId)));
    return (($self'owner, $self'count, $self'lastMessage, $self'lastSender, $self'lastReciver), ());
}

;;
;; Get methods of a Contract Proxy
;;

_ %owner() method_id(83229) {
    var self = $Proxy$_contract_load();
    var res = self~$Proxy$_fun_owner();
    return res;
}

_ %get_count() method_id(74564) {
    var self = $Proxy$_contract_load();
    var res = self~$Proxy$_fun_get_count();
    return res;
}

_ %get_last() method_id(92745) {
    var self = $Proxy$_contract_load();
    var res = self~$Proxy$_fun_get_last();
    return $LastMessage$_to_external(res);
}

_ supported_interfaces() method_id {
    return (
        "org.ton.introspection.v0"H >> 128,
        "org.ton.abi.ipfs.v0"H >> 128,
        "org.ton.deploy.lazy.v0"H >> 128,
        "org.ton.chain.workchain.v0"H >> 128,
        "org.ton.ownable"H >> 128
    );
}

_ get_abi_ipfs() method_id {
    return "ipfs://QmNaGZ3KbCDTK9RDEExsDDutxAtLjugHfV6XxUv1erbQpB";
}

_ lazy_deployment_completed() method_id {
    return get_data().begin_parse().load_int(1);
}

;;
;; Routing of a Contract Proxy
;;

((slice, int, slice, slice, slice), int) $Proxy$_contract_router_internal((slice, int, slice, slice, slice) self, int msg_bounced, slice in_msg) impure inline_ref {
    ;; Handle bounced messages
    if (msg_bounced) {
        return (self, true);
    }
    
    ;; Parse incoming message
    int op = 0;
    if (slice_bits(in_msg) >= 32) {
        op = in_msg.preload_uint(32);
    }
    
    
    ;; Receive ProxyMessage message
    if (op == 2376135367) {
        var msg = in_msg~$ProxyMessage$_load();
        self~$Proxy$_internal_binary_ProxyMessage(msg);
        return (self, true);
    }
    
    ;; Receive ChangeExampleOwner message
    if (op == 3797952890) {
        var msg = in_msg~$ChangeExampleOwner$_load();
        self~$Proxy$_internal_binary_ChangeExampleOwner(msg);
        return (self, true);
    }
    
    ;; Receive Deploy message
    if (op == 2490013878) {
        var msg = in_msg~$Deploy$_load();
        self~$Proxy$_internal_binary_Deploy(msg);
        return (self, true);
    }
    
    ;; Text Receivers
    if (op == 0) {
        var text_op = slice_hash(in_msg);
        
        ;; Receive "Repeat" message
        if (text_op == 0x561db9191cda1922faf6066ed24edc68166b3930bfeb59ba7fc844712e788f75) {
            self~$Proxy$_internal_text_561db9191cda1922faf6066ed24edc68166b3930bfeb59ba7fc844712e788f75();
            return (self, true);
        }
        
        ;; Receive "Return" message
        if (text_op == 0x6ef41a1603279cbd03396d3a3593d3f465c968554125a8b39067ebfc8b060d65) {
            self~$Proxy$_internal_text_6ef41a1603279cbd03396d3a3593d3f465c968554125a8b39067ebfc8b060d65();
            return (self, true);
        }
        if (slice_bits(in_msg) >= 32) {
            self~$Proxy$_internal_any_text(in_msg.skip_bits(32));
            return (self, true);
        }
    }
    
    return (self, false);
}

() recv_internal(int msg_value, cell in_msg_cell, slice in_msg) impure {
    
    ;; Context
    var cs = in_msg_cell.begin_parse();
    var msg_flags = cs~load_uint(4);
    var msg_bounced = -(msg_flags & 1);
    slice msg_sender_addr = __tact_verify_address(cs~load_msg_addr());
    __tact_context = (msg_bounced, msg_sender_addr, msg_value, cs);
    __tact_context_sender = msg_sender_addr;
    
    ;; Load contract data
    var self = $Proxy$_contract_load();
    
    ;; Handle operation
    int handled = self~$Proxy$_contract_router_internal(msg_bounced, in_msg);
    
    ;; Throw if not handled
    throw_unless(130, handled);
    
    ;; Persist state
    $Proxy$_contract_store(self);
}
