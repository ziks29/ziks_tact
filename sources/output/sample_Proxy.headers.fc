;;
;; Header files for Proxy
;; NOTE: declarations are sorted for optimal order
;;

;; __tact_verify_address
slice __tact_verify_address(slice address) inline;

;; __tact_load_address
(slice, slice) __tact_load_address(slice cs) inline;

;; __tact_load_address_opt
(slice, slice) __tact_load_address_opt(slice cs) inline;

;; __tact_store_address
builder __tact_store_address(builder b, slice address) inline;

;; __tact_store_address_opt
builder __tact_store_address_opt(builder b, slice address) inline;

;; __tact_my_balance
int __tact_my_balance() inline;

;; __tact_not_null
forall X -> X __tact_not_null(X x) inline;

;; __tact_context_get
(int, slice, int, slice) __tact_context_get() inline;

;; __tact_context_get_sender
slice __tact_context_get_sender() inline;

;; __tact_store_bool
builder __tact_store_bool(builder b, int v) inline;

;; __tact_slice_eq_bits
int __tact_slice_eq_bits(slice a, slice b) inline;

;; __tact_string_builder_start
tuple __tact_string_builder_start(builder b) inline;

;; __tact_string_builder_start_comment
tuple __tact_string_builder_start_comment() inline;

;; __tact_string_builder_end
cell __tact_string_builder_end(tuple builders) inline;

;; __tact_string_builder_append
((tuple), ()) __tact_string_builder_append(tuple builders, slice sc) inline_ref;

;; $Deploy$_load
(slice, ((int))) $Deploy$_load(slice sc_0) inline;

;; $DeployOk$_store
builder $DeployOk$_store(builder build_0, (int) v) inline;

;; $DeployOk$_store_cell
cell $DeployOk$_store_cell((int) v) inline;

;; $ProxyMessage$_load
(slice, ((slice, slice))) $ProxyMessage$_load(slice sc_0) inline;

;; $ChangeExampleOwner$_load
(slice, ((slice))) $ChangeExampleOwner$_load(slice sc_0) inline;

;; $Proxy$_store
builder $Proxy$_store(builder build_0, (slice, int, slice, slice, slice) v) inline;

;; $Proxy$_load
(slice, ((slice, int, slice, slice, slice))) $Proxy$_load(slice sc_0) inline;

;; $StateInit$_not_null
((cell, cell)) $StateInit$_not_null(tuple v) inline;

;; $LastMessage$_to_external
(slice, slice, slice) $LastMessage$_to_external(((slice, slice, slice)) v) inline;

;; $Proxy$init$_load
(slice, ((slice))) $Proxy$init$_load(slice sc_0) inline;

;; $Proxy$_contract_init
(slice, int, slice, slice, slice) $Proxy$_contract_init(slice $owner) impure inline_ref;

;; $Proxy$_contract_load
(slice, int, slice, slice, slice) $Proxy$_contract_load() impure inline_ref;

;; $Proxy$_contract_store
() $Proxy$_contract_store((slice, int, slice, slice, slice) v) impure inline;

;; $global_send
() $global_send((int, slice, int, int, cell, cell, cell) $params) impure inline_ref;

;; $String$_fun_asComment
cell $String$_fun_asComment(slice $self) impure inline_ref;

;; $Proxy$_fun_storeStats
((slice, int, slice, slice, slice), ()) $Proxy$_fun_storeStats((slice, int, slice, slice, slice) $self, slice $sender, slice $receiver, slice $message) impure inline_ref;

;; $SendParameters$_constructor_bounce_to_value_mode_body_code_data
((int, slice, int, int, cell, cell, cell)) $SendParameters$_constructor_bounce_to_value_mode_body_code_data(int bounce, slice to, int value, int mode, cell body, cell code, cell data) inline;

;; $Proxy$_fun_forward
((slice, int, slice, slice, slice), ()) $Proxy$_fun_forward((slice, int, slice, slice, slice) $self, slice $to, cell $body, int $bounce, tuple $init) impure inline_ref;

;; $Proxy$_fun_notify
((slice, int, slice, slice, slice), ()) $Proxy$_fun_notify((slice, int, slice, slice, slice) $self, cell $body) impure inline;

;; $Proxy$_fun_requireOwner
((slice, int, slice, slice, slice), ()) $Proxy$_fun_requireOwner((slice, int, slice, slice, slice) $self) impure inline_ref;

;; $Proxy$_fun_owner
((slice, int, slice, slice, slice), slice) $Proxy$_fun_owner((slice, int, slice, slice, slice) $self) impure inline_ref;

;; $Proxy$_fun_get_count
((slice, int, slice, slice, slice), int) $Proxy$_fun_get_count((slice, int, slice, slice, slice) $self) impure inline_ref;

;; $LastMessage$_constructor_lastMessage_lastSender_lastReciver
((slice, slice, slice)) $LastMessage$_constructor_lastMessage_lastSender_lastReciver(slice lastMessage, slice lastSender, slice lastReciver) inline;

;; $Proxy$_fun_get_last
((slice, int, slice, slice, slice), (slice, slice, slice)) $Proxy$_fun_get_last((slice, int, slice, slice, slice) $self) impure inline_ref;

;; $SendParameters$_constructor_to_value_mode_body
((int, slice, int, int, cell, cell, cell)) $SendParameters$_constructor_to_value_mode_body(slice to, int value, int mode, cell body) inline;

;; $DeployOk$_constructor_queryId
((int)) $DeployOk$_constructor_queryId(int queryId) inline;
