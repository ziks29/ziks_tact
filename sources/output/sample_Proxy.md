# TACT Compilation Report
Contract: Proxy
BOC Size: 1557 bytes

# Types
Total Types: 11

## StateInit
TLB: `_ code:^cell data:^cell = StateInit`
Signature: `StateInit{code:^cell,data:^cell}`

## Context
TLB: `_ bounced:bool sender:address value:int257 raw:^slice = Context`
Signature: `Context{bounced:bool,sender:address,value:int257,raw:^slice}`

## SendParameters
TLB: `_ bounce:bool to:address value:int257 mode:int257 body:Maybe ^cell code:Maybe ^cell data:Maybe ^cell = SendParameters`
Signature: `SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell}`

## ChangeOwner
TLB: `change_owner#819dbe99 queryId:uint64 newOwner:address = ChangeOwner`
Signature: `ChangeOwner{queryId:uint64,newOwner:address}`

## ChangeOwnerOk
TLB: `change_owner_ok#327b2b4a queryId:uint64 newOwner:address = ChangeOwnerOk`
Signature: `ChangeOwnerOk{queryId:uint64,newOwner:address}`

## Deploy
TLB: `deploy#946a98b6 queryId:uint64 = Deploy`
Signature: `Deploy{queryId:uint64}`

## DeployOk
TLB: `deploy_ok#aff90f57 queryId:uint64 = DeployOk`
Signature: `DeployOk{queryId:uint64}`

## FactoryDeploy
TLB: `factory_deploy#6d0ff13b queryId:uint64 cashback:address = FactoryDeploy`
Signature: `FactoryDeploy{queryId:uint64,cashback:address}`

## ProxyMessage
TLB: `proxy_message#8da0f2c7 str:^string to:address = ProxyMessage`
Signature: `ProxyMessage{str:^string,to:address}`

## ChangeExampleOwner
TLB: `change_example_owner#e260297a newOwner:address = ChangeExampleOwner`
Signature: `ChangeExampleOwner{newOwner:address}`

## LastMessage
TLB: `_ lastMessage:Maybe ^string lastSender:Maybe address lastReciver:Maybe address = LastMessage`
Signature: `LastMessage{lastMessage:Maybe ^string,lastSender:Maybe address,lastReciver:Maybe address}`

# Get Methods
Total Get Methods: 3

## owner

## get_count

## get_last

# Error Codes
2: Stack undeflow
3: Stack overflow
4: Integer overflow
5: Integer out of expected range
6: Invalid opcode
7: Type check error
8: Cell overflow
9: Cell underflow
10: Dictionary error
13: Out of gas error
32: Method ID not found
34: Action is invalid or not supported
37: Not enough TON
38: Not enough extra-currencies
128: Null reference exception
129: Invalid serialization prefix
130: Invalid incoming message
131: Constraints error
132: Access denied
133: Contract stopped
134: Invalid argument
135: Code of a contract was not found
136: Invalid address
137: Masterchain support is not enabled for this contract
27831: Only owner can call this function