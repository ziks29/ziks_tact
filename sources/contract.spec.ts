import { toNano } from "@ton/core";
import { ContractSystem } from "@tact-lang/emulator";
import { Proxy } from "./output/sample_Proxy";
import { inspect } from "util";

describe("contract", () => {
    it("should forward messages", async () => {
        // Create ContractSystem and deploy contract
        let system = await ContractSystem.create();
        let owner = system.treasure("owner");
        let user1 = system.treasure("user1");
        let user2 = system.treasure("user2");
        let proxy = system.open(await Proxy.fromInit(owner.address));
        let proxy_tracker = system.track(proxy);
        await proxy.send(user1, { value: toNano("0.1") }, { $$type: "ProxyMessage", str: "Hello", to: user2.address });
        await system.run();
        expect(await proxy.getGetCount()).toBe(1n);
        expect((await proxy.getGetLast()).lastMessage).toEqual("Hello");
        expect((await proxy.getGetLast()).lastReciver?.equals(user2.address)).toBeTruthy();
        expect((await proxy.getGetLast()).lastSender?.equals(user1.address)).toBeTruthy();
        let proxy_events = proxy_tracker.collect();
        expect(proxy_events).toMatchSnapshot();
    });
});
