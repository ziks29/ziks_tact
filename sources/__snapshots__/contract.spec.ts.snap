// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contract should forward messages 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{8DA0F2C78005B941F64C273187231080074EA967FB2824F7202A134E1378F236E89BBF7508B_}
 x{48656C6C6F}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(user1)",
          "to": "kQCnn6u31i9SSh4gZaaeRZlS9dkOfltRnRW5R14VDfzDmE4R",
          "type": "internal",
          "value": "0.1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 13051n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Hello",
              "type": "text",
            },
            "bounce": true,
            "from": "kQCnn6u31i9SSh4gZaaeRZlS9dkOfltRnRW5R14VDfzDmE4R",
            "to": "@treasure(user2)",
            "type": "internal",
            "value": "0.085777",
          },
        ],
      },
    ],
  },
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "cell": "x{FFFFFFFF0000000048656C6C6F}",
            "type": "cell",
          },
          "bounce": false,
          "from": "@treasure(user2)",
          "to": "kQCnn6u31i9SSh4gZaaeRZlS9dkOfltRnRW5R14VDfzDmE4R",
          "type": "internal",
          "value": "0.084777",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8088n,
      },
    ],
  },
]
`;
